runtime/vm/jnimisc.cpp

```diff
@@ -168,6 +168,37 @@ getCurrentClassLoader(J9VMThread *currentThread)
 		} else {
 			/* use the sender (native method) classloader */
 			classLoader = J9_CLASS_FROM_METHOD(nativeMethod)->classLoader;
+#if JAVA_SPEC_VERSION >= 15
+			/* special case - if jdk/internal/loader/NativeLibraries.load(NativeLibraryImpl impl, String name, boolean isBuiltin, boolean isJNI)
+			 * is the current native method, use the class loader of "impl.fromClass".
+			 * This nativeMethod can't be cached cause HCR might make it invalid.
+			 */
+			if (J9VMJDKINTERNALLOADERNATIVELIBRARIES_LOAD_METHOD(currentThread->javaVM) == nativeMethod) {
+				J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
+				vmFuncs->internalEnterVMFromJNI(currentThread);
+				/* The current native method has a NativeLibraryImpl instance as its first argument */
+				j9object_t nativeLibraryImplObject = (j9object_t)(currentThread->arg0EA[0]);
+				Trc_VM_findNativeLibrariesLoad_nativeMethod(currentThread, nativeMethod, nativeLibraryImplObject, classLoader);
+				/* Handle the object reference being redirected by the stack grower */
+				if (J9_ARE_ANY_BITS_SET((UDATA)nativeLibraryImplObject, 1)) {
+					nativeLibraryImplObject = *(j9object_t*)((UDATA)nativeLibraryImplObject - 1);
+				}
+				if (NULL != nativeLibraryImplObject) {
```

Comment:
Can this every really be null?

Reply:
It shouldn't, change it to `Assert_VM_notNull(nativeLibraryImplObject)`.

Reply:
Is it the receiver of a virtual call? If so, it can't be NULL, if not, it technically could be, so the if is fine.